solr/solrj/src/java/org/apache/solr/common/util/CallRouter.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.common.util;
+
+public interface CallRouter {
+    /**
+     * send to a specific node. usually admin requests
+     */
+    CallRouter toNode(String nodeName);
+
+    /**
+     * Make a request to any replica of the shard of type
+     */
+    CallRouter toShard(String collection, String shard, ReplicaType type);
+
+    /**
+     * Identify the shard using the route key and send the request to a given replica type
+     */
+    CallRouter toShard(String collection, ReplicaType type, String routeKey);
```

Comment:
can we reorder this so the decision maker for the routing i.e. routeKey is the 2nd param like `CallRouter toShard(String collection, String shard, ReplicaType type);` ?

Reply:
it would conflict with the other method . we can't say if the second param is `shard` or `routeKey`

Reply:
Ah yes. I misread the first signature as (String, String, String).